test(cli): new cli test suite

lalvarezt opened the pull request:
new set of test that cover the overall functionality of the cli

lalvarezt commented:
@alexpasmantier can you help me with the test issues, seems the bat command is not installed (workflow config I assume) and there's a timing problem with the pty because the preview is not loaded (workflow), locally works just fine

alexpasmantier commented:
Noticed a couple of flaky tests which I fixed in the last commit.

lalvarezt commented:
> Noticed a couple of flaky tests which I fixed in the last commit.

thanks, those were patches to try to get rid of the pesky workflow issue 😅
I don't have anything else for this one, unless you notice something you want to tweak

alexpasmantier commented on television/config/keybindings.rs:
```diff
@@ -31,14 +32,76 @@ impl Display for Binding {
+                A: MapAccess<'de>,
+            {
+                let mut bindings = FxHashMap::default();
+
+                while let Some(key) = map.next_key::<String>()? {
+                    let binding = map.next_value::<Binding>()?;
+
+                    // Handle special toggle feature keys
+                    let action = match key.as_str() {
+                        "toggle_remote_control" => {
+                            Action::ToggleFeature(FeatureFlags::RemoteControl)
+                        }
+                        "toggle_preview" => {
+                            Action::ToggleFeature(FeatureFlags::PreviewPanel)
+                        }
+                        "toggle_help" => {
+                            Action::ToggleFeature(FeatureFlags::HelpPanel)
+                        }
+                        "toggle_status_bar" => {
+                            Action::ToggleFeature(FeatureFlags::StatusBar)
+                        }
+                        _ => {
+                            // Try to deserialize as regular Action enum using TOML value
+                            let action_str = toml::Value::String(key.clone());
+                            Action::deserialize(action_str)
+                                .map_err(serde::de::Error::custom)?
+                        }
+                    };
+
+                    bindings.insert(action, binding);
+                }
+
+                Ok(KeyBindings(bindings))
+            }
+        }
+
+        deserializer.deserialize_map(KeyBindingsVisitor)
+    }
+}
+
```
I feel like this shouldn't be done here but should be handled when processing actions instead, e.g. keybindings keeps the same derived deserialize but then when a user hits a key that sends `Action::ToggleRemoteControl`, it's processed as it if were `Action::ToggleFeature(FeatureFlags::RemoteControl)` and we can isolate that retro compatibility logic in a dedicated function:


```rust
fn handle_action(action: &Action) {
    match action {
        ...
        Action::ToggleRemoteControl | Action::ToggleHelp | ... => handle_legacy_actions(...),
    }
}
```

lalvarezt replied:
I played a lot with what you are suggesting and always ended up in a place where I had to add a lot of code to support the legacy options (without the de-serializer I mean).

Since the old syntax (toggle_remote_control, toggle_preview, …) is purely an input-format concern, I thought that we could just catch it early and convert it to the new logic. 

Also this shouldn't affect the end-user because their config will remain the same, just the developers 🤔

what do you think? perhaps you have a different idea or see a simple way of doing that doesn't affect performance?

alexpasmantier replied:
Ok, just pushed a commit that I believe makes things a little less coupled and a bit clearer. wdyt?

lalvarezt replied:
looks good, I'll update the doc I was writing on the features feature to reflect this approach instead. 
Inadvertently you removed a check that prevented the toggling of the preview when the remote was open. Just open remote and toggle preview, you'll see it changes but the hints are not updated

alexpasmantier replied:
Ah yeah ok, I was reading the diff over and over not being able to find what you were referring to 😅

alexpasmantier commented:
Hmmm, not sure why that test is flaky 🤔

alexpasmantier commented:
Ok, looks good, tell me when you're ready to merge this 😊

lalvarezt commented:
> Ok, looks good, tell me when you're ready to merge this 😊

ok, fixed the regression and added a test to catch it in the future.
Added new doc for features, please have a look. Other than that it's ok for me

alexpasmantier commented:
> Added new doc for features, please have a look.

Here's my honest opinion:

I don't know if this was done using AI (if it wasn't you're one hell of a productive writer, congrats 😅) but either way the phrasing really makes it feel like it was ai-generated and I think we should change the style to avoid setting off any potential contributors.

I also think the documentation is too verbose and beats too much around the bush.

Example:

> Television Features System
> The Television Features System is a UI component management framework that provides dual-state control over interface elements. This document provides an in-depth analysis of the features functionality, its architecture, purpose, and configuration.
>
> Television's Features System enables granular control over UI components through a dual-state paradigm. Each feature can be independently controlled along two dimensions:
> 
>     Enabled/Disabled: Whether the feature's functionality is available
>     Visible/Hidden: Whether the feature is displayed in the interface
> 
> This design pattern allows for UI management where features can exist in three meaningful states: Active (enabled and visible), Hidden (enabled but not visible), and Disabled (completely inactive).
> Key Benefits
> 
>     - Flexible UI Control: Users can customize their interface without losing functionality
>     - Runtime Toggleability: Features can be dynamically shown/hidden during application use
>     - Configuration Persistence: Feature states defined in configuration files persist across sessions
>     - CLI Override Support: Command-line flags can override configured feature states for individual sessions
>     - Channel-Specific Defaults: Different channels can have different feature configurations


I'm thinking something more straight to the point like:

> Features
> Television manages UI component states using a system of _features_ that dictate which components of the application should be shown/hidden/enabled/disabled at any given moment.

which I believe conveys in the end almost as much info for any developer reading this.

I wouldn't insist too much on "how this design pattern is great", listing its benefits etc. and would remove that entirely.

I think the module analysis section should actually live next to the code (as docstrings) and doesn't have a lot of added value here without being able to navigate the code.

The `Feature Components` and `State Management` sections have nice visuals which we should keep but I would remove documentation that merely lists the methods and properties available (little added value in this context, and the added value is in the details like whether state manipulation methods always succeed or sometimes fail etc. which adds a lot to the overall maintenance surface of the project and I feel should live inside docstrings directly in the code).

I think `Advanced Usage` in the table of contents doesn't have any associated content.

TLDR: 
If I had to rewrite it, I'd make it much shorter, more straight to the point, would add docstrings if necessary and keep the visual diagrams 👍🏻

alexpasmantier commented:
Thanks for the PR and the updates 🙏🏻, let's merge this 🎉
